### PR TITLE
feat: syft attestor

### PR DESCRIPTION
**Pull request by mikhailswift**
Adds an attestor that creates a SBOM using syft for a provided source or
an automatically detected image tar file from the context's products.

Signed-off-by: Mikhail Swift <mikhail@testifysec.com>

### CODE REVIEW

**Comment by colek42** [comment]
the digest here is md5

```
                    "path": "/usr/lib/apache2/modules/mod_authz_user.so",
                    "digest": {
                      "algorithm": "md5",
                      "value": "a68a61bba6b829f044fd1e01acfcabdc"
                    },
                    "isConfigFile": false
```

is there a way to use whatever is in the attestation context.  

from the docs it looks like it should be configurable.

```  # the file digest algorithms to use when cataloging files (options: "sha256", "md5", "sha1")
  # SYFT_FILE_METADATA_DIGESTS env var
  digests: ["sha256"]
```

**Comment by colek42** [comment]
- Looks like there is a vuln introduced with syft.  Can we pin gin to a recent version?
- We are introducing MPL and GPL code here.  I think we need to document it in our repository.  There will also be considerations for any downstream consumers.

**Comment by mikhailswift** [comment]
I'm doubtful that there is actually GPL'd code being compiled into Witness here.  The first instance snyk finds is 

` github.com/testifysec/witness@0.0.0 › github.com/opencontainers/image-spec@v1.0.3-0.20220114050600-8b9d41f48198 › github.com/golangci/golangci-lint@v1.43.0 › github.com/OpenPeeDeeP/depguard@v1.0.1 `

open-containers/image-spec is licensed under Apache 2, the same as Witness and does not import golangci-lint anywhere in their go code, so we're not compiling in ours, either.  I'll do a more thorough look through the rest but it seems like snyk is picking up image-spec's CI tools as part of the library.

**Comment by mikhailswift** [comment]
I also don't believe we're vulnerable to either of the vulnerabilities detected.

the gin vulnerability is focused around gin's logging functionality -- which we're not using at all

the opa vuln we're also not using -- we're using an unaffected version of opa for our rego evaluation.

**Comment by mikhailswift** [comment]
I did try replacing gin with a newer version, seemed to compile.

**Comment by mikhailswift** [comment]
go mod graph is not showing any versions of opa below 0.37 which doesn't have that CVE.

go mod graph is not showing rapid as being part of our project -- it seems like a test library so I'm guessing some dependency uses it in their tests.

anchore/go-version is MPL.  MPL isn't viral like GPL and is only concerned with modifications to the MPL code -- which we're not doing.

**Comment by mikhailswift** [comment]
I'll look into the hashing stuff in a bit.

**Comment by jeff-mccoy** [comment]
👋  what's the plan for this PR to be merged in?  We have a PR in [Zarf](https://zarf.dev) pointing to a commit hash in this PR.  Thanks!

**Comment by mikhailswift** [comment]
This will be merged in today

File: go.mod [hunk 1]
```diff
@@ -111,3 +162,7 @@ require (
 )
 
 replace github.com/sigstore/rekor => github.com/testifysec/rekor v0.4.0-dsse-intermediates-2
+
+replace github.com/gin-gonic/gin v1.5.0 => github.com/gin-gonic/gin v1.7.7
+
+replace github.com/opencontainers/image-spec => github.com/opencontainers/image-spec v1.0.3-0.20220303224323-02efb9a75ee1
```
Comment (colek42) [inline go.mod:168]: is there a reason for this?

**Review by colek42 — APPROVED** [review]

**Comment by colek42** [comment]
I have some issues about the dependencies we are pulling in here -- this is a good short term fix, however, we need to bring SBOM functionality in-tree

**Comment by mikhailswift** [comment]
> I have some issues about the dependencies we are pulling in here -- this is a good short term fix, however, we need to bring SBOM functionality in-tree

yes as we've discussed.

**Pull request merged**